crates/services/txpool_v2/src/pending_pool.rs

```diff
@@ -0,0 +1,519 @@
+                        .into_iter()
+                        .filter(|input| match input {
+                            MissingInput::Utxo(utxo) => *utxo != utxo_id,
+                            _ => true,
+                        })
+                        .collect();
+                    if missing_inputs.is_empty() {
+                        resolved_txs.push(tx);
+                    } else {
+                        self.unresolved_inputs_by_tx
+                            .insert(tx_id, (tx, missing_inputs));
+                    }
+                }
+            }
+        }
+    }
+
+    fn new_known_contract(
+        &mut self,
+        contract_id: ContractId,
+        resolved_txs: &mut Vec<(ArcPoolTx, InsertionSource)>,
+    ) {
+        if let Some(tx_ids) = self
+            .unresolved_txs_by_inputs
+            .remove(&MissingInput::Contract(contract_id))
+        {
+            for tx_id in tx_ids {
+                if let Some((tx, missing_inputs)) =
+                    self.unresolved_inputs_by_tx.remove(&tx_id)
+                {
+                    let missing_inputs: Vec<MissingInput> = missing_inputs
+                        .into_iter()
+                        .filter(|input| match input {
+                            MissingInput::Contract(contract) => *contract != contract_id,
+                            _ => true,
+                        })
+                        .collect();
+                    if missing_inputs.is_empty() {
+                        resolved_txs.push(tx);
+                    } else {
+                        self.unresolved_inputs_by_tx
+                            .insert(tx_id, (tx, missing_inputs));
+                    }
+                }
+            }
+        }
+    }
+
+    fn expire_transactions(
+        &mut self,
+        expired_txs: &mut Vec<(ArcPoolTx, InsertionSource, Vec<MissingInput>)>,
+    ) {
+        let now = SystemTime::now();
+        while let Some((ttl, tx_id)) = self.ttl_check.back() {
+            if *ttl > now {
+                break;
+            }
+            if let Some(((tx, source), missing_inputs)) =
+                self.unresolved_inputs_by_tx.remove(tx_id)
+            {
+                self.current_bytes =
+                    self.current_bytes.saturating_sub(tx.metered_bytes_size());
+                self.current_gas = self.current_gas.saturating_sub(tx.max_gas());
+                self.current_txs = self.current_txs.saturating_sub(1);
+                for input in &missing_inputs {
+                    // Remove tx_id from the list of unresolved transactions for this input
+                    // If the list becomes empty, remove the entry
+                    if let Some(tx_ids) = self.unresolved_txs_by_inputs.get_mut(input) {
+                        tx_ids.retain(|id| *id != *tx_id);
+                        if tx_ids.is_empty() {
+                            self.unresolved_txs_by_inputs.remove(input);
+                        }
```

Comment:
If you used `entry`, then you don't need do search in the `self.unresolved_txs_by_inputs` to times

Reply:
I don't see how I can do the deletion of the entry if tx_ids is empty directly with the entry API. Can you provide an example ?

Reply:
Done.